client/bqms_run/ksh.py

```diff
@@ -0,0 +1,228 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Processors for heredoc SQL embedded in KSH code to translate."""
+
+import logging
+import re
+from enum import Enum
+from typing import List, Optional
+
+
+logger = logging.getLogger(__name__)
+
+
+class ShellFragmentType(Enum):
+    SHELL = "SHELL"
+    HEREDOC = "HEREDOC"  # aka bteq
+    EMPTY = "EMPTY"  # a special one
+
+
+class ShellFragment:
+    """
+    A fragment of a KSH file.
+    """
+
+    def __init__(
+        self, text: str,
+        line: int,
+        fragment_type: ShellFragmentType,
+        quoted: bool
+    ) -> None:
+        super().__init__()
+        self.text = text
+        self.line = line
+        self.fragment_type = fragment_type
+        self.quoted = quoted
+
+    def __str__(self) -> str:
+        return (
+            "ShellFragment{"
+            + ",".join(
+                [
+                    f"line={self.line}",
+                    f"fragment_type={self.fragment_type}",
+                    f"quoted={self.quoted}",
+                    f"text={self.text}",
+                ]
+            )
+            + "}"
+        )
+
+    def get_fragment_type(self) -> ShellFragmentType:
+        return self.fragment_type
+
+    def get_text(self) -> str:
+        return self.text
+
+
+class ExtractorState(Enum):
+    SHELL = "SHELL"
+    BTEQ = "BTEQ"
+
+
+class KshExtractor:
```

Comment:
It seems this is a direct port of the java version, correct? Comments and warnings and all :-)

Reply:
Yes